### ISSUE TITLE
Feedback

### ISSUE THREAD

**Issue by krishnadey30** [Rapiz1/chapel-heap#7]
Overall it looks good to me. I liked the idea of having parallel safe operations. Tests are also good. 

----

What is the use of `boundsChecking`? I didn't find the definition of this variable anywhere. You can give some more information about it.

https://github.com/Rapiz1/chapel-heap/blob/531fca9fa5d47dd78a1afbf5fb661642f92a955e/src/Heap.chpl#L257

----

Isn't it a helper function for comparison? I am not sure if `_greater` is a correct method name for this one. (correct me if I am wrong) 
https://github.com/Rapiz1/chapel-heap/blob/531fca9fa5d47dd78a1afbf5fb661642f92a955e/src/Heap.chpl#L269

----

I would suggest naming it as `_heapify_down` and another one to `_heapify_up`.(Just a suggestion).

https://github.com/Rapiz1/chapel-heap/blob/531fca9fa5d47dd78a1afbf5fb661642f92a955e/src/Heap.chpl#L291

----

I would like to see some other static methods such as:
- `make_heap`
- `pop_heap`
- `push_heap`

**Comment by yujqiao** [comment]
Thanks a lot for your valuable feedback! It will help me a lot!
1. `boundsChecking` comes from the standard library `HaltWrappers`. I learned this when I was browsing the source of `List`
2. It is a helper function for comparison. I name it `_greater` because I want to imply that we're implementing a max-heap, which reduces the mental burden of considering comparator. Also, it's consistent with that when `DefaultComparator` is passed, a max-heap will be created.
3. The names you give sounds great! I will rename them immediately.
4. That's a good idea. I will implement them soon.

Also, I have been trying hard to figure out why this `init=` won't work. With this, even a single `var h1 = new heap(int, ReverseComparator);` can fail to compile. However, without this function, it can pass.
https://github.com/Rapiz1/chapel-heap/blob/2564b7abb71df648b0d5c6d5f81dcd95e4c03328/src/Heap.chpl#L121-L139
```
Updating mason-registry
$CHPL_HOME/modules/internal/ChapelBase.chpl:1633: error: unresolved call 'heap(int(64),ReverseComparator,false).init=(heap(int(64),ReverseComparator,false))'
/home/rapiz/Heap/src/Heap.chpl:128: note: this candidate did not match: heap.init=(other: heap)
$CHPL_HOME/modules/internal/ChapelBase.chpl:1633: note: because method call receiver with type heap(int(64),ReverseComparator,false)
/home/rapiz/Heap/src/Heap.chpl:67: note: is passed to formal 'ref this: heap(int(64),ReverseComparator,false)'
compilation failed for minimalExample.chpl
```
Here is `minimalExample.chpl`:
```
Use Heap;
Use Sort;
var h1 = new heap(int, ReverseComparator);
```

**Comment by yujqiao** [comment]
@krishnadey30 Can you give me some clue?

**Comment by krishnadey30** [comment]
```chpl
    proc init=(other: heap) {
      if !isCopyableType(other.eltType) then
        compilerError("Cannot copy heap with element type that cannot be copied");

      this.eltType = other.eltType;
      this.comparator = other.comparator;
      this.parSafe = other.parSafe;
      this.complete();
      _commonInitFromIterable(other._data);
    }
```

This worked perfectly for me.

**Comment by yujqiao** [comment]
That looks good!

**Comment by krishnadey30** [comment]
> 1. `boundsChecking` comes from the standard library `HaltWrappers`. I learned this when I was browsing the source of `List`

AFAIK `boundChecking` is a compile-time constant. Let me check once and I will let you know about it.

**Comment by yujqiao** [comment]
> ```chapel
>     proc init=(other: heap) {
>       if !isCopyableType(other.eltType) then
>         compilerError("Cannot copy heap with element type that cannot be copied");
> 
>       this.eltType = other.eltType;
>       this.comparator = other.comparator;
>       this.parSafe = other.parSafe;
>       this.complete();
>       _commonInitFromIterable(other._data);
>     }
> ```
> 
> This worked perfectly for me.

It's very strange that the compilation still fails with the same error on my machine. I use the chapel from master branch.

**Comment by yujqiao** [comment]
Do you think it's a bug or so? The error message makes no sense to me ( or maybe I don't really understand) .

**Comment by yujqiao** [comment]
@krishnadey30 Hey, I just add `makeHeap` and `pushHeap` as you said in https://github.com/Rapiz1/chapel-heap/commit/ddac78d4f4d7ae3bd6b98aeb4ec71ac3f5d4341d along with tests and docs. This module feels much more complete now! 
Also, I have renamed the functions in https://github.com/Rapiz1/chapel-heap/commit/d3da9cd17c773cf2567c8c89265d68cb56ad1ef6

**Comment by cassella** [comment]
I get the same compilation failure with both versions of the `init=`.

I think it's something to do with the comparator.  I made a version of Heap.chpl that doesn't use the comparator, (`_greater()` just returns `x > y`)and the minimal example (with no comparator argument) compiles against it.

And the minimal example works with your Heap.chpl with that `init=` enabled, if the comparator is changed from `ReverseComparator` to `DefaultComparator`.

I think it might have to do with `ReverseComparator` being a generic type -- it has a `comparator` field since it can be used to reverse any other comparator, not just the `DefaultComparator`.

But when I make the example pass `ReverseComparator(DefaultComparator)` and have that `init=` enabled, I get a different error.

```
Heap.chpl:268: In function '_greater':
Heap.chpl:270: error: A type alias of 'ReverseComparator' may not be used in a new-expression because it contains a typeless field ('comparator')
Heap.chpl:297: Function '_greater' instantiated as: _greater(this: heap(int(64),ReverseComparator(DefaultComparator),false))
```

But when I comment out that `init=`, that same example works.


I think this is some kind of subtlety where when your record's `comparator` type field is set to `ReverseComparator(DefaultComparator)`, it's an instantiation of a generic type, and then `_greater()` invoking `new comparator()` is getting into trouble.   I think that's worth some discussion in gitter, because this isn't very obvious to me.

You might also want to open an issue requesting a better error message for your `init=` and example as written, because that error message is not at all helpful in figuring this out, as you've probably noticed.  :)   I narrowed it down to the comparator only by comparing to List.chpl (since your code is based on that) and seeing only the comparator as structurally different.


For the purposes of the `Heap`, I'd explore making your record contain a variable that's an instance of the comparator passed in, and not a type alias of it.  (Or make the user pass in an actual object instead of a type when creating the `Heap`.  Note that the code in `Sort.chpl` passes instances, not types: `defaultComparator`, not `DefaultComparator`.)

**Comment by cassella** [comment]
```chapel
 /*
      Build the heap from elements that have been stored, from bottom to top
      in O(N)
    */
    pragma "no doc"
    proc _commonInitFromIterable(iterable) {
```

Are you sure this is O(n)?

**Comment by yujqiao** [comment]
> ```chapel
>  /*
>       Build the heap from elements that have been stored, from bottom to top
>       in O(N)
>     */
>     pragma "no doc"
>     proc _commonInitFromIterable(iterable) {
> ```
> 
> Are you sure this is O(n)?

I think so. I don't know how to prove it but I have heard about it. And some searching leads me to this: https://www.geeksforgeeks.org/time-complexity-of-building-a-heap/

**Comment by cassella** [comment]
Oh neat.  I'd forgotten about that.

**Comment by yujqiao** [comment]
@cassella I have read your comment https://github.com/Rapiz1/chapel-heap/issues/7#issuecomment-606354585 Very helpful! Thanks!

1. I will try to come up with a minimal example to reproduce the error message and open an issue.
2. I have just asked about `new ReverseComparator(DefaultComparator)` in Gitter.
3. Passing an instance of comparator sounds good. However, in C++, `std::sort` needs an instance and classes need a type as their template argument, not an argument of the constructor. I will do more research on why they do that and think about how should we handle this.

**Comment by cassella** [comment]
I still don't know what's going on exactly with the combination of the `type compator` and the `proc init=()`.

But even apart from that, having the `Heap` keep an instance instead of a type feels more natural to me.  (What if the user wants to use a comparator containing a regular int or bool flag?  You wouldn't be able to instantiate it from the type without knowing those arguments.  Like this silly example,

```chapel
record closerToN {
  var N;
  proc init(N) { this.N = N; }
  proc compare(a, b) {
    var ad = abs(a - N), bd = abs(b - N);
    if (ad < bd) { return -1; }
    else if (bd < ad) { return 1; }
    else return 0;
  }
}
```

A user could set up a comparator-instance heap with a `new closerToN(5)`, but if they set up a comparator-type heap with `closerToN`, how can you have the `5` available for when you actually need to instantiate one?

Between that, and the `init=` error doing it this way, and Sort's precedent, I think the instance-based heap is the way to go.

> However, in C++, std::sort needs an instance and classes need a type as their template argument, not an argument of the constructor.

I'm not following.  Is this a problem you're stuck on?

**Comment by yujqiao** [comment]
@cassella Yes, having an instance is more flexible. And it's easier to implement, saving lots of troubles. So I'm going to use an instance.

> I'm not following. Is this a problem you're stuck on?

In C++, `std::priority_queue` needs a type parameter used as a comparator to get specialized. 
And `std::sort` accepts an instance as the argument.
```
int main() {
	priority_queue<int, vector<int>, greater<int>> q; // greater<int> is a type
	vector<int> v = {1,2,3};
	sort(v.begin(), v.end(), greater<int>()); // greater<int>() is an instance
}
```
I failed to find out why they design like that. We're facing a similiar design issue here. However, I think passing a instance is just fine in our case.
